### PR TITLE
feat(trace): Add header and upgrade TransactionTrace

### PR DESCRIPTION
**Pull request by mask-pp**
Description:
- Change `BlockResult`:
> Delete origin `BlockTrace`.
> Add `Coinbase *AccountWrapper`、`Header  *Header`、`Transactions []*TransactionData`
> Rename from `BlockResult` to `BlockTrace`

- Upgrade `TransactionTrace`, use `type.Transaction` structure in order to avoid unnecessary fields copy:
> fields change:
>> data => input
>> txHash => hash

> Old struct:
```
{
	"type": 0,
	"nonce": 0,
	"txHash": "0xb2febc1213baec968f6575789108e175273b8da8f412468098893084229f1542",
	"gas": 500000,
	"gasPrice": "0x3b9aec2e",
	"from": "0x1c5a77d9fa7ef466951b2f01f724bca3a5820b63",
	"to": "0xc0c4c8baea3f6acb49b6e1fb9e2adeceeacb0ca2",
	"chainId": "0xcf55",
	"value": "0x152d02c7e14af6000000",
	"data": "0x",
	"isCreate": false,
	"v": "0x19ece",
	"r": "0xab07ae99c67aa78e7ba5cf6781e90cc32b219b1de102513d56548a41e86df514",
	"s": "0x34cbd19feacd73e8ce64d00c4d1996b9b5243c578fd7f51bfaec288bbaf42a8b"	
}
```
> New struct:
```
{
        "type": "0x2",
        "nonce": "0x2",
        "gasPrice": null,
        "maxPriorityFeePerGas": "0x3b9ace45",
        "maxFeePerGas": "0x3b9af65b",
        "gas": "0x11ce2",
        "value": "0x0",
        "input": "0x40c10f1900000000000000000000000001bae6bf68e9a03fb2bc0615b1bf0d69ce9411ed00000000000000000000000000000000000000000000003635c9adc5dea00000",
        "v": "0x0",
        "r": "0x2879c4f495eb1f6b7a36f39a8912d7d568f140e5221ddf81288cd311ce495c3d",
        "s": "0x711b9fdaec2fab37f0bff89a723cc7854b973e4968b1a5a0365851cfd8e7f574",
        "to": "0xc671842470ceaea4cd0341f405a846b4dd8faf7c",
        "chainId": "0xcf55",
        "accessList": [],
        "hash": "0xccbe20bb215b71d4659569caf379412a9344e0de39cc5339d75164650f683814"
}
```

### CODE REVIEW

File: core/types/l2trace_block.go [hunk 1]
```diff
@@ -20,20 +21,10 @@ type BlockTrace struct {
...
-	Value    *hexutil.Big    `json:"value"`
-	Data     string          `json:"data"`
-	IsCreate bool            `json:"isCreate"`
-	V        *hexutil.Big    `json:"v"`
-	R        *hexutil.Big    `json:"r"`
-	S        *hexutil.Big    `json:"s"`
+	ChainId   *hexutil.Big   `json:"chainId"`
+	IsCreate  bool           `json:"isCreate"`
+	From      common.Address `json:"from"`
+	TxContent *Transaction   `json:"txContent"`
```
Comment (0xmountaintop) [inline core/types/l2trace_block.go]:
```suggestion
	*Transaction
```

Comment (0xmountaintop) [reply core/types/l2trace_block.go]: so that we can keep those fields in the same level

**Comment by 0xmountaintop** [comment]
This is a breaking change, so it's not a "refactor"

File: core/types/l2trace_block.go [hunk 2]
```diff
@@ -4,78 +4,25 @@ import (
...
-	Number       *hexutil.Big        `json:"number"`
-	Hash         common.Hash         `json:"hash"`
-	GasLimit     uint64              `json:"gasLimit"`
-	Difficulty   *hexutil.Big        `json:"difficulty"`
-	BaseFee      *hexutil.Big        `json:"baseFee"`
-	Coinbase     *AccountWrapper     `json:"coinbase"`
-	Time         uint64              `json:"time"`
-	Transactions []*TransactionTrace `json:"transactions"`
+type TransactionData struct {
+	ChainId  *big.Int       `json:"chainId"`
```
Comment (0xmountaintop) [inline core/types/l2trace_block.go]: I don't think we need this. looks like `*Transaction` already contains it

File: eth/tracers/api_blockresult.go [hunk 3]
```diff
@@ -19,7 +19,7 @@ import (
 )
 
 type TraceBlock interface {
-	GetBlockResultByNumberOrHash(ctx context.Context, blockNrOrHash rpc.BlockNumberOrHash, config *TraceConfig) (trace *types.BlockResult, err error)
+	GetBlockResultByNumberOrHash(ctx context.Context, blockNrOrHash rpc.BlockNumberOrHash, config *TraceConfig) (trace *types.BlockTrace, err error)
```
Comment (0xmountaintop) [inline eth/tracers/api_blockresult.go]:
```suggestion
	GetBlockTraceByNumberOrHash(ctx context.Context, blockNrOrHash rpc.BlockNumberOrHash, config *TraceConfig) (trace *types.BlockTrace, err error)
```

File: eth/tracers/api_blockresult.go [hunk 4]
```diff
@@ -42,8 +42,8 @@ type traceEnv struct {
 	executionResults []*types.ExecutionResult
 }
 
-// GetBlockResultByNumberOrHash replays the block and returns the structured BlockResult by hash or number.
-func (api *API) GetBlockResultByNumberOrHash(ctx context.Context, blockNrOrHash rpc.BlockNumberOrHash, config *TraceConfig) (trace *types.BlockResult, err error) {
+// GetBlockResultByNumberOrHash replays the block and returns the structured BlockTrace by hash or number.
+func (api *API) GetBlockResultByNumberOrHash(ctx context.Context, blockNrOrHash rpc.BlockNumberOrHash, config *TraceConfig) (trace *types.BlockTrace, err error) {
```
Comment (0xmountaintop) [inline eth/tracers/api_blockresult.go]:
```suggestion
// GetBlockTraceByNumberOrHash replays the block and returns the structured BlockTrace by hash or number.
func (api *API) GetBlockTraceByNumberOrHash(ctx context.Context, blockNrOrHash rpc.BlockNumberOrHash, config *TraceConfig) (trace *types.BlockTrace, err error) {
```

File: eth/tracers/api_blockresult.go [hunk 5]
```diff
@@ -130,7 +130,7 @@ func (api *API) createTraceEnv(ctx context.Context, config *TraceConfig, block *
 	return env, nil
 }
 
-func (api *API) getBlockResult(block *types.Block, env *traceEnv) (*types.BlockResult, error) {
+func (api *API) getBlockResult(block *types.Block, env *traceEnv) (*types.BlockTrace, error) {
```
Comment (0xmountaintop) [inline eth/tracers/api_blockresult.go]:
```suggestion
func (api *API) getBlockTrace(block *types.Block, env *traceEnv) (*types.BlockTrace, error) {
```

File: eth/tracers/api_blockresult.go [hunk 6]
```diff
@@ -329,21 +329,28 @@ func (api *API) getTxResult(env *traceEnv, state *state.StateDB, index int, bloc
 }
 
 // Fill blockResult content after all the txs are finished running.
-func (api *API) fillBlockResult(env *traceEnv, block *types.Block) (*types.BlockResult, error) {
+func (api *API) fillBlockResult(env *traceEnv, block *types.Block) (*types.BlockTrace, error) {
```
Comment (0xmountaintop) [inline eth/tracers/api_blockresult.go]:
```suggestion
func (api *API) fillBlockTrace(env *traceEnv, block *types.Block) (*types.BlockTrace, error) {
```

File: eth/tracers/api_blockresult.go [hunk 7]
```diff
@@ -329,21 +329,28 @@ func (api *API) getTxResult(env *traceEnv, state *state.StateDB, index int, bloc
...
-		CodeHash: statedb.GetCodeHash(env.coinbase),
+
+	txs := make([]*types.TransactionData, block.Transactions().Len())
+	for i, tx := range block.Transactions() {
+		txs[i] = types.NewTraceTransaction(tx, block.NumberU64(), api.backend.ChainConfig())
 	}
-	blockResult := &types.BlockResult{
-		BlockTrace:       types.NewTraceBlock(api.backend.ChainConfig(), block, &coinbase),
+
+	blockResult := &types.BlockTrace{
```
Comment (0xmountaintop) [inline eth/tracers/api_blockresult.go]:
```suggestion
	blockTrace := &types.BlockTrace{
```

File: eth/tracers/api_blockresult_test.go [hunk 8]
```diff
@@ -100,7 +100,7 @@ func TestAPI_GetBlockResultByNumberOrHash(t *testing.T) {
 	checkChainAndProof(t, backend, parent, block, blockResult)
 
 	// check txs.
-	checkTxs(t, block.Transactions(), blockResult.BlockTrace.Transactions)
+	checkTxs(t, block.Transactions(), blockResult.Transactions)
```
Comment (0xmountaintop) [inline eth/tracers/api_blockresult_test.go]:
```suggestion
	checkTxs(t, block.Transactions(), blockTrace.Transactions)
```

File: eth/tracers/api_blockresult_test.go [hunk 9]
```diff
@@ -109,7 +109,7 @@ func TestAPI_GetBlockResultByNumberOrHash(t *testing.T) {
 	checkStructLogs(t, txTraces, blockResult.ExecutionResults)
 
 	// check coinbase
-	checkCoinbase(t, backend, blockResult.BlockTrace.Coinbase)
+	checkCoinbase(t, backend, blockResult.Coinbase)
```
Comment (0xmountaintop) [inline eth/tracers/api_blockresult_test.go]:
```suggestion
	checkCoinbase(t, backend, blockTrace.Coinbase)
```

File: eth/tracers/api_blockresult_test.go [hunk 10]
```diff
@@ -109,7 +109,7 @@ func TestAPI_GetBlockResultByNumberOrHash(t *testing.T) {
 	checkStructLogs(t, txTraces, blockResult.ExecutionResults)
```
Comment (0xmountaintop) [inline eth/tracers/api_blockresult_test.go]:
```suggestion
	checkStructLogs(t, txTraces, blockTrace.ExecutionResults)
```

File: eth/tracers/api_blockresult_test.go [hunk 11]
```diff
@@ -120,7 +120,7 @@ func verifyProof(t *testing.T, expect [][]byte, actual []hexutil.Bytes) {
 	}
 }
 
-func checkChainAndProof(t *testing.T, b *testBackend, parent *types.Block, block *types.Block, blockResult *types.BlockResult) {
+func checkChainAndProof(t *testing.T, b *testBackend, parent *types.Block, block *types.Block, blockResult *types.BlockTrace) {
```
Comment (0xmountaintop) [inline eth/tracers/api_blockresult_test.go]:
```suggestion
func checkChainAndProof(t *testing.T, b *testBackend, parent *types.Block, block *types.Block, blockTrace *types.BlockTrace) {
```

File: eth/tracers/api_blockresult_test.go [hunk 12]
```diff
@@ -130,8 +130,8 @@ func checkChainAndProof(t *testing.T, b *testBackend, parent *types.Block, block
 	assert.NoError(t, err)
 
 	storageProof := blockResult.StorageTrace.StorageProofs
-	for _, tx := range blockResult.BlockTrace.Transactions {
-		for _, addr := range []common.Address{tx.From, *tx.To} {
+	for _, tx := range blockResult.Transactions {
```
Comment (0xmountaintop) [inline eth/tracers/api_blockresult_test.go]:
```suggestion
	for _, tx := range blockTrace.Transactions {
```

File: eth/tracers/api_blockresult_test.go [hunk 13]
```diff
@@ -120,7 +120,7 @@ func verifyProof(t *testing.T, expect [][]byte, actual []hexutil.Bytes) {
 	}
 }
 
-func checkChainAndProof(t *testing.T, b *testBackend, parent *types.Block, block *types.Block, blockResult *types.BlockResult) {
+func checkChainAndProof(t *testing.T, b *testBackend, parent *types.Block, block *types.Block, blockResult *types.BlockTrace) {
 	assert.Equal(t, parent.Hash(), block.ParentHash())
 	storgeTrace := blockResult.StorageTrace
```
Comment (0xmountaintop) [inline eth/tracers/api_blockresult_test.go]:
```suggestion
	storgeTrace := blockTrace.StorageTrace
```

File: eth/tracers/api_blockresult_test.go [hunk 14]
```diff
@@ -130,8 +130,8 @@ func checkChainAndProof(t *testing.T, b *testBackend, parent *types.Block, block
 	assert.NoError(t, err)
 
 	storageProof := blockResult.StorageTrace.StorageProofs
```
Comment (0xmountaintop) [inline eth/tracers/api_blockresult_test.go]:
```suggestion
	storageProof := blockTrace.StorageTrace.StorageProofs
```

**Review by 0xmountaintop — CHANGES_REQUESTED** [review]
rename all the original blockResult to blockTrace

**Comment by 0xmountaintop** [comment]
1. we should recover eth/downloader/downloader.go & les/downloader/downloader.go
2. rename eth/tracers/api_blockresult.go -> eth/tracers/api_blocktrace.go, eth/tracers/api_blockresult_test.go -> eth/tracers/api_blocktrace_test.go

**Comment by mask-pp** [comment]
> 1. we should recover eth/downloader/downloader.go & les/downloader/downloader.go
> 2. rename eth/tracers/api_blockresult.go -> eth/tracers/api_blocktrace.go, eth/tracers/api_blockresult_test.go -> eth/tracers/api_blocktrace_test.go

changed.

**Review by 0xmountaintop — APPROVED** [review]

**Review by Lawliet-Chan — APPROVED** [review]
LGTM, and will change common-rs/types/eth.rs and scroll/common/message/message.go after merging this PR.

**Review by scroll-dev — APPROVED** [review]

**Review by noel2004 — COMMENTED** [review]
With `coinbase` field being reserved, mpt witness generator should be not affected and the update is LGTM.

Also we should note that the `MPTWitness` (and maybe also `Coinbase`) fields would be deprecated once the mpt-witness generator in common-rs had worked.

File: core/types/l2trace.go [hunk 15]
```diff
@@ -21,9 +21,11 @@ var (
 	}
 )
 
-// BlockResult contains block execution traces and results required for rollers.
-type BlockResult struct {
-	BlockTrace       *BlockTrace        `json:"blockTrace"`
+// BlockTrace contains block execution traces and results required for rollers.
+type BlockTrace struct {
+	Coinbase         *AccountWrapper    `json:"coinbase"`
```
Comment (icemelon) [inline core/types/l2trace.go:26]: One quick question. Why do we add an extra `Coinbase` here given the `Header` already contains the `Coinbase`?

File: core/types/l2trace.go [hunk 16]
```diff
@@ -21,9 +21,11 @@ var (
...
+	Header           *Header            `json:"header"`
```
Comment (icemelon) [inline core/types/l2trace.go:27]: we need to add a list of block hashes for `BLOCKHASH` opcode to use, starting from the current block.

Comment (mask-pp) [reply core/types/l2trace.go:27]: The `BLOCKHASH` opcode is supported by this [func](https://github.com/scroll-tech/go-ethereum/blob/0bdd2e0575a2b5741b757a9966120ef83ecddbad/core/evm.go#L57).

File: core/types/l2trace_block.go [hunk 17]
```diff
@@ -4,78 +4,23 @@ import (
...
-	Hash         common.Hash         `json:"hash"`
-	GasLimit     uint64              `json:"gasLimit"`
-	Difficulty   *hexutil.Big        `json:"difficulty"`
-	BaseFee      *hexutil.Big        `json:"baseFee"`
-	Coinbase     *AccountWrapper     `json:"coinbase"`
-	Time         uint64              `json:"time"`
-	Transactions []*TransactionTrace `json:"transactions"`
+type TransactionData struct {
+	IsCreate bool           `json:"isCreate"`
+	From     common.Address `json:"from"`
```
Comment (Lawliet-Chan) [inline core/types/l2trace_block.go:12]: I think the `From` should be `json omitempty`, isn't it?

**Pull request merged**